design/one-pager-controlled-rollout-of-composition-functions-v2.md

```diff
@@ -0,0 +1,198 @@
+it (see the package manager section below). If a `FunctionRuntime` already
+exists, the composition revision will be added as an owner reference.
+
+A user wishing to roll out a new version of `function-patch-and-transform`
+simply updates their composition to reference the new version, resulting in
+creation of a new composition revision and runtime creation for the new function
+version. The existing `compositionUpdatePolicy` and
+`compositionRevisionSelector` mechanisms on XRs can be used to progressively
+roll out the change:
+
+```yaml
+apiVersion: custom-api.example.org/v1alpha1
+kind: AcmeBucket
+metadata:
+  name: my-bucket
+spec:
+  compositionUpdatePolicy: Manual
+  compositionRevisionSelector:
+    matchLabels:
+      release-channel: alpha
+```
+
+In addition to allowing for progressive rollout, this new way of specifying
+functions is simpler and safer to use than the existing method. Users don't need
+to specify all their functions as dependencies or otherwise pre-install
+them. They also don't need to know how the package manager names dependency
+packages to construct the right `functionRef.name` in their composition. They
+are also guaranteed to get exactly the version they specify regardless of what
+other changes happen in the cluster; pacakges can be specified by digest for
+further safety.
+
+### Package Manager Changes
+
+Currently, both the package revision controller and the package runtime
+controller operate on package revision resources (`FunctionRevision` and
+`ProviderRevision`). The revision controller installs resources (CRDs, etc.)
+from the package, while the runtime controller creates runtime resources
+(deployments, services, etc.) for active revisions of packages that have a
+runtime component.
+
+In order to allow runtime-only packages, we will introduce new `FunctionRuntime`
+and `ProviderRuntime` resources, which will be reconciled by the package runtime
+controller. The revision controller will be changed to create a runtime resource
+corresponding to the active revision; the runtime controller will no longer
+operate on revisions directly.
+
+This allows the composition revision controller to create its own
+`FunctionRuntime` resources to run functions without creating `Function`s or
+`FunctionRevision`s. The package revision and composition revision controllers
+will use the same scheme to name their `FunctionRevision`s, ensuring that a
+particular version of a function runs only once even if it's installed multiple
+ways. I.e., a `FunctionRuntime` may be shared between a `FunctionRevision` and
+one or more `CompositionRevision`s.
+
+Note that this change has two side effects for functions installed by the
+composition revision controller:
+
+1. Their dependencies will not be installed, as there will be no revision
+   inserted into the `Lock`.
+2. They will not have corresponding package or package revision resources,
+   meaning they will not show up in `kubectl get pkg` or similar commands.
```

Comment:
I have mixed feelings about them not showing up in `kubectl get pkg`.

I like this design a lot overall, but I worry about the asymmetry we're introducing. Functions are still a package in the artifact sense, but no longer a package in the API object sense. If you run `kubectl get pkg` you're not really seeing all the packages that're "installed" on your control plane anymore.